### PR TITLE
Teach dao

### PR DESCRIPTION
**Pull request by taylorjbrown**
This is code for the CSV file uploader. The HTML needs to be rendered for this code to run. A basic HTML page to test is in the repository as well as a Template that works correct. This also needs to be added to the upload HTML page.

### CODE REVIEW

File: src/api/upload.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import resource from 'resource-router-middleware'
+
+
+'use strict';
```
Comment (jamesvclements) [inline src/api/upload.js]: not needed

File: src/api/upload.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+var formidable = require('formidable');
```
Comment (jamesvclements) [inline src/api/upload.js]: use `import` instead of `require`

File: src/api/upload.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+//var csv_parser = require('./csvParser');
+var parse_csv = require('../services/csvUpload/testing_csvjson');
+
+
+export default ({ config, db }) => resource({
+	
+	/** POST - Create a new upload */
+	create(req, res) {
+
+		var form = new formidable.IncomingForm();
```
Comment (jamesvclements) [inline src/api/upload.js]: `let` instead of `var`

File: src/api/upload.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    
+    	form.multiples = true;
+    	form.uploadDir = path.join(__dirname, '../services/csvUpload/uploads');
+        console.log(form.uploadDir);
```
Comment (jamesvclements) [inline src/api/upload.js]: remove logs used while testing

File: src/api/upload.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+export default ({ config, db }) => resource({
+	
+	/** POST - Create a new upload */
+	create(req, res) {
+
+		var form = new formidable.IncomingForm();
+    
+    	form.multiples = true;
+    	form.uploadDir = path.join(__dirname, '../services/csvUpload/uploads');
```
Comment (jamesvclements) [inline src/api/upload.js]: no semicolons - see README for linting to find and fix all linting issues

File: src/api/upload.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        console.log(form.uploadDir);
+      	form.on('file', function rename_file(field, file) {
+        	fs.renameSync(file.path, path.join(form.uploadDir, file.name));
+            parse_csv(null,path.join(form.uploadDir, file.name));
+    	});
+
+    	form.on('error', function(err) {
+        	console.log('An error has occured: \n' + err);
```
Comment (jamesvclements) [inline src/api/upload.js]: use `console.error` when logging errors

File: src/api/upload.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import resource from 'resource-router-middleware'
+
+
+'use strict';
+
+var formidable = require('formidable');
+var path = require('path');
+var fs = require('fs');
+//var csv_parser = require('./csvParser');
+var parse_csv = require('../services/csvUpload/testing_csvjson');
```
Comment (jamesvclements) [inline src/api/upload.js]: use camelCase

File: src/api/upload.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    	form.on('end', function() {
+        	res.end('success');
+    	});
+    	form.parse(req);
+		}
+})
+
+/*
+function pass_csv(form, cb) {
```
Comment (jamesvclements) [inline src/api/upload.js]: remove unused functions

File: src/config.json [hunk 9]
```diff
@@ -1,5 +1,8 @@
 {
 	"port": 8080,
 	"bodyLimit": "100kb",
-	"corsHeaders": ["Link"]
+	"corsHeaders": ["Link"],
+    "mysql_user": "root",
```
Comment (jamesvclements) [inline src/config.json]: Belongs in `.env` file - no secrets in the config

File: src/gradeCSVTemplate.csv [hunk 10]
```diff
@@ -0,0 +1 @@
+year,id,lastName,firstName,sex,dob,gradeEntering,homeroomTeacher,asp,nextMeetingSch,advancedMath,speechLanguage,studentDevelopment,mathEnrichment,IUreadingServices,IUmathServices,earobics,behavior,workEthic,youngestChild,onlyChild,newStudent,medicalConcern,hmp,dra,RAZ,WTW,iStation,mathBench,Dibels,cogAT,IOWA,elaTotal,ExtendedELA,mathTotal,facultyStudent,potentialDelay,behaviorObservation,selfHelp,socialEmotional,dial4,comments
```
Comment (jamesvclements) [inline src/gradeCSVTemplate.csv]: Create a subdirectory under `src/services` for uploads (`src/services/upload`) and put any related files in there.

File: src/load_csv_data.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
+'use strict';
```
Comment (jamesvclements) [inline src/load_csv_data.js]: Remove `'use strict'`. Also, file-naming convention uses hyphens so call this file `load-csv-data.js`. This would also go in that `services` subdirectory I mentioned

File: src/load_csv_data.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+var extend = require('lodash').assign;
+var mysql = require('mysql');
+var config = require('../config');
+
+
+function getConnection () {
```
Comment (jamesvclements) [inline src/load_csv_data.js]: We're using dependency injection, so rather than creating the connection here it should be passed into this file instead. See the way its being done through the api to the DAOs

File: src/load_csv_data.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  return mysql.createConnection(extend({
+    database: 'STM_DB'
+  }, {
+    host: config.get('MYSQL_HOST'),
+    user: config.get('MYSQL_USER'),
+    password: config.get('MYSQL_PASSWORD')
+  }));
+}
+
+function insert_student (data, cb) {
```
Comment (jamesvclements) [inline src/load_csv_data.js]: camelCase

File: src/services/csvUpload/load_csv_data.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
+'use strict';
```
Comment (jamesvclements) [inline src/services/csvUpload/load_csv_data.js]: Okay, so you did create this directory - nice. Just remove all the old files then

File: src/services/csvUpload/load_csv_data.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  connection.end();
+}
+
+
+module.exports = {
+    insert_student: insert_student,
+    insert_ydsd: insert_ydsd
+};
+
+if (module == require.main) {
```
Comment (jamesvclements) [inline src/services/csvUpload/load_csv_data.js]: I'm not entirely sure what this does, but I don't think you need to run depending on the module. It'll be run when the service is called from the api

File: src/services/csvUpload/load_csv_data.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+      data3,
+      cb
+    );
+
+  connection.end();
+}
+
+
+module.exports = {
```
Comment (jamesvclements) [inline src/services/csvUpload/load_csv_data.js]: Just export the functions you want right in front of them, see other files for examples.

File: src/services/csvUpload/uploads/gradeCSVTemplate.csv [hunk 17]
```diff
@@ -0,0 +1 @@
...
NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,Yes,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL
NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,Yes,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL
NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,Yes,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL
NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,Yes,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL
NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,Yes,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL
```
Comment (jamesvclements) [inline src/services/csvUpload/uploads/gradeCSVTemplate.csv]: Just out of curiousity - what do all the nulls do for the template?

File: src/uploadCSV.html [hunk 18]
```diff
@@ -0,0 +1,30 @@
+<html>
```
Comment (jamesvclements) [inline src/uploadCSV.html]: Just send us this file via Slack - it doesn't belong in the server repo

File: src/services/csvUpload/testing_csvjson.js [hunk 19]
```diff
@@ -0,0 +1,37 @@
+
```
Comment (jamesvclements) [inline src/services/csvUpload/testing_csvjson.js]: If this is a test file, rename it `csvjson.spec.js`. All files with the `.spec.js` extension will be run on `npm test`.

File: src/services/csvUpload/uploads/gradeCSVTemplate.csv (hunk 17 above)
Comment (taylorjbrown) [reply src/services/csvUpload/uploads/gradeCSVTemplate.csv]: This was originally for the DB, but after more testing, empty strings are allowed so in the next push these NULLs will not be there.

File: src/services/csvUpload/testing_csvjson.js (hunk 19 above)
Comment (taylorjbrown) [reply src/services/csvUpload/testing_csvjson.js]: This was just bad naming convention. This is an actual file needed to run the code and is not for testing and will be name .

**Comment by taylorjbrown** [comment]
I have updated the files to the comments that you made above James. Let me know if there is anything else that needs to be corrected before it can be pushed to master.

**Comment by jamesvclements** [comment]
This is being redone in the newDB branch
